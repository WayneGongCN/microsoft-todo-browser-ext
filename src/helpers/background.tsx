import { BackgroundContext } from '../../types';
import { ErrorCode } from '../constants/enums';
import AppError from './error';
import { timing } from './report';

export let backgroundContext: BackgroundContext;

const loadBackground = () =>
  new Promise((resolve, reject) => {
    chrome.runtime.getBackgroundPage((ctx: Window & { backgroundContext: BackgroundContext }) => {
      timing('background loaded', performance.now())
      const lastError = chrome.runtime.lastError;
      if (lastError) reject(new AppError({ code: ErrorCode.UNKNOW, message: lastError?.message }));
      else {
        backgroundContext = ctx.backgroundContext
        resolve(backgroundContext);
      }
    });
  });

/**
 * 获取 Background 页面的上下文
 */
export const getBackgroundCtx = async () => {
  return backgroundContext || await loadBackground();
};
