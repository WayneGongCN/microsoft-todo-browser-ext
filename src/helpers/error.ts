import { SerializError } from '../../types';

export enum ErrorCode {
  UNKNOW = -1,
  NOT_FOUND_ACCOUNT,
  ACQUIRE_TOKEN,
  ACQUIRE_TOKEN_SILENT,
}

const errorMessageMap = {
  [ErrorCode.UNKNOW]: '',
  [ErrorCode.NOT_FOUND_ACCOUNT]: '',
  [ErrorCode.ACQUIRE_TOKEN]: '',
  [ErrorCode.ACQUIRE_TOKEN_SILENT]: '',
};

export default class AppError extends Error {
  code: ErrorCode;
  time: number;

  /**
   * 优先使用传入的 error message
   */
  constructor({ code, message }: { code: ErrorCode; message?: string }) {
    super(message || errorMessageMap[code]);

    this.time = Date.now();
    this.code = code;
  }

  serializ(): SerializError {
    return {
      code: this.code,
      // 优先使用已定义过的 error message
      message: errorMessageMap[this.code] || this.message,
      stack: this.stack,
      time: this.time,
    };
  }
}
