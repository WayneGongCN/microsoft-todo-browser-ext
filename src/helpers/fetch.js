export default (tokenOptions) => (url, method, data, options) => {
  const action = getOAuthToken(tokenOptions);
  if (method === 'GET') options = data;

  return window.store.dispatch(action)
    .then((res) => res.accessToken)
    .then((token) => fetch(url, {
      method,
      body: JSON.stringify(data),
      headers: { Authorization: `Bearer ${token}`, 'Content-Type': 'application/json' },
      ...options,
    }))
    // eslint-disable-next-line no-nested-ternary
    .then((res) => (res.ok ? res.status === 200 ? res.json() : res : Promise.reject(res)))
    .then(res => {
      console.log(res)
      return res
    })
};