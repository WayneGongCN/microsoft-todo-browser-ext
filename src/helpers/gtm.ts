/* eslint-disable @typescript-eslint/ban-ts-comment */
export default (i = process.env.GTM_ID, w = window, d = document, s = 'script', l = 'dataLayer') => {
  // @ts-ignore
  w[l] = w[l] || [];
  // @ts-ignore
  w[l].push({ 'gtm.start': new Date().getTime(), event: 'gtm.js' });
  const f = d.getElementsByTagName(s)[0],
    j = d.createElement(s),
    dl = l != 'dataLayer' ? '&l=' + l : '';
  // @ts-ignore
  j.async = true;
  // @ts-ignore
  j.src = 'https://www.googletagmanager.com/gtm.js?id=' + i + dl + process.env.GTM_ENV;
  // @ts-ignore
  f.parentNode.insertBefore(j, f);
};
