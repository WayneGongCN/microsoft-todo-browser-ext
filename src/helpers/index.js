import Notify from './notification';


export const isDev = process.env.NODE_ENV === 'development';
export const isProd = process.env.NODE_ENV === 'production';


export const triggerByHotkey = (fn, hotkey = 'Enter') => (e) => {
  if (e.type === 'keyup' && e.code !== hotkey) return false;
  return fn(e);
};

export const getEventValue = (fn, key = 'value') => (e) => fn(e.target[key]);

export const openMicrosoftTodo = (target) => new Promise((resolve, reject) => {
  const prefix = 'https://to-do.live.com';

  let url = '';
  // eslint-disable-next-line no-undef
  if (target instanceof Task) {
    url = `${prefix}/tasks/id/${target.id}/details`;
    // eslint-disable-next-line no-undef
  } else if (target instanceof Tasklist) {
    url = `${prefix}/tasks/${target.id}`;
  } else {
    url = `${prefix}/tasks/inbox`;
  }

  chrome.tabs.create({ active: true, url }, (tab) => {
    if (chrome.runtime.lastError) return reject(chrome.runtime.lastError.message);
    return resolve(tab);
  });
});


export const makeBookmarkInfo = () => new Promise((resolve, reject) => {
  chrome.tabs.query({ active: true }, (tab) => {
    if (chrome.runtime.lastError) return reject(chrome.runtime.lastError.message);
    const curTab = tab[0] || '';
    return resolve(curTab && `\n\n---\n${curTab.title}\n${curTab.url}`);
  });
});


export const showNavigateNotify = (target, title = 'Notify', message = 'Open Microsoft To-Do.') => {
  const notify = new Notify(title, message);
  notify.onClick(() => {
    notify.clear();
    openMicrosoftTodo(target);
  });
  notify.create();
  return notify;
};

export const sendMessageToActiveTab = (msg) => new Promise((resolve, reject) => {
  chrome.tabs.query({ active: true }, (res) => {
    if (!res.length) reject(new Error('Not found active tab.'));
    const activeTab = res[0];
    chrome.tabs.sendMessage(activeTab.id, msg, (response) => {
      console.log('response: ', response);
    });
  });
});
