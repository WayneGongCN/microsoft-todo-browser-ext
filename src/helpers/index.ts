import { Task, Tasklist } from '../models/Task';
import Notify from './notification';


export const isDev = process.env.NODE_ENV === 'development';
export const isProd = process.env.NODE_ENV === 'production';


export const openMicrosoftTodo = (target?: Task | Tasklist) => new Promise((resolve, reject) => {
  const prefix = 'https://to-do.live.com';

  let url = '';
  // eslint-disable-next-line no-undef
  if (target instanceof Task) {
    url = `${prefix}/tasks/id/${target.id}/details`;
    // eslint-disable-next-line no-undef
  } else if (target instanceof Tasklist) {
    url = `${prefix}/tasks/${target.id}`;
  } else {
    url = `${prefix}/tasks/inbox`;
  }

  chrome.tabs.create({ active: true, url }, (tab) => {
    if (chrome.runtime.lastError) return reject(chrome.runtime.lastError.message);
    return resolve(tab);
  });
});


export const getActiveTab = () => new Promise((resolve, reject) => {
  chrome.tabs.query({ active: true }, (tabs) => {
    if (chrome.runtime.lastError) reject(chrome.runtime.lastError.message);
    else if (tabs && tabs.length > 0) {
      const firstActiveTab = tabs[0];
      resolve(firstActiveTab);
    } else {
      reject(new Error('Not found active tab.'));
    }
  });
});


export const makeBookmarkInfo = (tab, quickAdd = false) => {
  const bookmarkInfo = `\n${tab.title}\n\n${tab.url}\n`;
  const quickAddInfo = "\nCreated by 'Quick add task'\n";
  return `\n\n---${quickAdd ? quickAddInfo : ''}${bookmarkInfo}`;
};


export const sendMessageToActiveTab = (msg) => getActiveTab().then((tab) => chrome.tabs.sendMessage(tab.id, msg));


export const showNotify = (title = 'Notify', message) => {
  const notify = new Notify(title, message);
  notify.create();
  return notify;
};


export const showTaskNotify = (task, title, message = 'Open Microsoft To-Do.') => {
  const notify = showNotify(title, message);
  notify.onClick(() => {
    notify.clear();
    openMicrosoftTodo(task);
  });
  return notify;
};

