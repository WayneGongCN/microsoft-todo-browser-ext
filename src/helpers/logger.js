const backgroundPage = chrome.extension.getBackgroundPage();
const logger = backgroundPage.console;

export default logger;
