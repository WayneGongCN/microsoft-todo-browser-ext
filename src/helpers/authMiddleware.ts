import { State } from "../redux";
import { authentication } from "../redux/auth";
import { logger } from "./logger";

let authPromise: Promise<void> = null;

const authMiddleware = (store: any) => (next: any) => (action: any) => {
  // logger.log("authMiddleware", action.type);
  const { type, meta, payload } = action;
  if (!type) return next(action);
  if (type.includes("authenticationResult")) return next(action);

  const state: State = store.getState();
  let accessToken = state.auth.authenticationResult?.accessToken;

  if (accessToken) {
    // logger.log('已有 token', action.type)
    action.meta = {
      ...meta,
      accessToken,
    };
    return next(action);
  } else {
    if (authPromise) {
      // logger.log('正在获取 token 拦截 action', action.type)
      authPromise.then(() => {
        // logger.log('token 获取成功1，放过 action', action.type)
        next(action);
      });
    } else {
      authPromise = store.dispatch(authentication())
      // logger.log('开始获取 token', action.type, authPromise)
      
      authPromise.then(() => {
        // logger.log('token 获取成功2，放过 action', action.type)
        next(action)
      });
    }
  }
};

export default authMiddleware;
