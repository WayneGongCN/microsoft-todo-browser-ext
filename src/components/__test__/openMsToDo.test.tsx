/* eslint-disable @typescript-eslint/ban-ts-comment */
import React from 'react';
import snapshotRender from 'react-test-renderer';
import { openMicrosoftTodo } from '../../helpers/index';
import { cleanup, fireEvent, screen, render as domRender } from '@testing-library/react';
import OpenMSTodo from '../OpenMSToDo';


describe('Snapshot test', () => {
  test('normal', () => {
    const tree = snapshotRender.create(<OpenMSTodo />).toJSON();
    expect(tree).toMatchSnapshot('normal');
  });
});


jest.mock('../../helpers/index', () => ({ openMicrosoftTodo: jest.fn() }));
describe('Event test', () => {
  afterEach(cleanup);

  test('Click link', () => {
    domRender(<OpenMSTodo />);

    const link = screen.getByText('Microsoft To Do');
    fireEvent.click(link);

    // @ts-ignore
    expect(openMicrosoftTodo.mock.calls.length).toEqual(1);
    // @ts-ignore
    expect(openMicrosoftTodo.mock.calls[0][0]).toBe(undefined)
  });
});
