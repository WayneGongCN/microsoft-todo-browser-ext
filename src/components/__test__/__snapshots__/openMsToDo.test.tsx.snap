// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test normal: normal 1`] = `
<button
  className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineNone MuiLink-button css-93ncls-MuiTypography-root-MuiLink-root"
  id="com-btn-open-ms"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
>
  Microsoft To Do
</button>
`;
